global/scripts/debug.shlib

```diff
@@ -0,0 +1,275 @@
+#!/bin/echo This script should not run directly:
+
+# Description:
+#
+#   Function for printing debugging information on caught errors.
+#   It makes use of error trapping. 
+#
+#   To disable error trapping run `debug_disable_trap`
+#   To disable exit uppon trapped error run `debug_disable_exit`
+
+#
+# Input:
+#
+#   $1 - line number
+#   $2 - command ran
+#   $3 - exit status
+
+trap 'debug_on_error $LINENO "$BASH_COMMAND" $?' ERR
+
+if [[ "$(type -t log_Err)" != "function" ]]
+then
+    source "$HCPPIPEDIR/global/scripts/log.shlib" "$@"
+fi
+
+HCP_DEBUG_USETRAP=ON
+HCP_DEBUG_EXIT_ON_ERROR=ON
+
+if [ -z ${HCP_DEBUG_COLOR} ] ; then
+    HCP_DEBUG_COLOR=FALSE
+fi
+
+if [ -z ${HCP_VERBOSE} ] ; then
+    HCP_VERBOSE=FALSE
+fi
+
+# Description:
+#
+#   Function to turn on error trapping
+
+debug_enable_trap()
+{
+    HCP_DEBUG_USETRAP=ON
+}
+
+# Description:
+#
+#   Function to turn off error trapping
+
+debug_disable_trap()
+{
+    HCP_DEBUG_USETRAP=OFF
+}
+
+# Description:
+#
+#   Function to turn on exit upon error trap
+
+debug_enable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=ON
+}
+
+# Description:
+#
+#   Function to turn off exit upon error trap
+
+debug_disable_exit()
+{
+    HCP_DEBUG_EXIT_ON_ERROR=OFF
+}
+
+# Description:
```

Comment:
Can all this coloring related stuff come later, so that we get the definition of `debug_on_error()` (which is the "core" of this shlib) near the top?

Reply:
It uses `debug_red_echo` so that has to be defined first. I put all the rest after it.

Reply:
Functions do not need to be defined before defining other functions that use them, they only need to be defined before the functions or code that use them are called/run.  The debug_on_err function can come first.

Reply:
Thanks @coalsont -- that was exactly the question I was going to ask.

Reply:
Ok. Will change.

Reply:
To take this further, we can define a defaults function, which can call the enable functions instead of using the magic variable names and values, and then call it at the end of the script:

```
<top of file, shebang>
<comments>
debug_on_error()
{
...
}

debug_defaults()
{
    debug_enable_trap
    debug_enable_exit
    #the ${<var>:-} pattern exists to work properly under set -eu, mac's old bash doesn't have [[ -v <var> ]]
    if [[ "${HCP_DEBUG_COLOR:-}" == "" ]]
    then
        debug_disable_color
    fi
    if [[ "${HCP_VERBOSE:-}" == "" ]]
    then
        debug_disable_verbose
    fi
}
...
<other functions>
debug_defaults
trap 'debug_on_error "$LINENO" "$BASH_COMMAND" "$?"' ERR
```

Reply:
Done.